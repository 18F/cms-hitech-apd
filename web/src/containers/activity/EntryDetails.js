import { Button, Review } from '@cmsgov/design-system-core';
import PropTypes from 'prop-types';
import React, { Fragment, useMemo, useRef, useState } from 'react';

import { Provider } from './ActivityContext';
import ContractorResources from './ContractorResources';
import CostAllocate from './CostAllocate';
import Costs from './Costs';
import Overview from './Overview';
import Goals from './Goals';
import Schedule from './Schedule';
import StandardsAndConditions from './StandardsAndConditions';
import { TimesCircle } from '../../components/Icons';
import { t } from '../../i18n';

const makeTitle = ({ name, fundingSource }, i) => {
  let title = `${t('activities.namePrefix')} ${i}`;
  if (name) {
    title += `: ${name}`;
  }
  if (fundingSource) {
    title += ` (${fundingSource})`;
  }
  return title;
};

const EntryDetails = ({ activity, index }) => {
  const container = useRef();

  const aKey = activity.key;
  const num = index + 1;

  const [collapsed, internalSetCollapsed] = useState(num > 1);
  const setCollapsed = newCollapsed => {
    if (newCollapsed) {
      const { top } = container.current.getBoundingClientRect();
      if (top < 0 || top > window.innerHeight) {
        container.current.scrollIntoView({ behavior: 'auto' });
        container.current.focus();
      }
    }
    internalSetCollapsed(newCollapsed);
  };

  const title = useMemo(() => makeTitle(activity, num), [
    activity.fundingSource,
    activity.name,
    num
  ]);

  const editContent = useMemo(
    () => (
      <div className="nowrap visibility--screen">
        <Button
          size="small"
          variation="transparent"
          onClick={() => setCollapsed(!collapsed)}
        >
          {collapsed ? (
            'Edit'
          ) : (
            <Fragment>
              <TimesCircle /> Close
            </Fragment>
          )}
        </Button>
      </div>
    ),
    [collapsed]
  );

  return (
    <Provider value={{ activity, index }}>
      <div
        id={`activity-${aKey}`}
        className={`activity--body activity--body__${
          collapsed ? 'collapsed' : 'expanded'
        } activity--body__${num === 1 ? 'first' : 'notfirst'}`}
        ref={container}
      >
        <Review heading={title} headingLevel={4} editContent={editContent} />
        <div className={collapsed ? 'visibility--print' : ''}>
          <Overview aKey={aKey} />
          <Goals aKey={aKey} />
          <Schedule aKey={aKey} />
          <Costs aKey={aKey} />
          <ContractorResources aKey={aKey} />
          <CostAllocate aKey={aKey} />
          <StandardsAndConditions aKey={aKey} />
          <Button variation="primary" onClick={() => setCollapsed(true)}>
            Done
          </Button>
        </div>
      </div>
    </Provider>
  );
};

EntryDetails.propTypes = {
  activity: PropTypes.object.isRequired,
  index: PropTypes.number.isRequired
};

export default EntryDetails;
