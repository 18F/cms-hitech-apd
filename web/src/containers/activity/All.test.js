import { mount } from 'enzyme';
import React from 'react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import Activities from './All';

const mockStore = configureStore([thunk]);

// Mock out the EntryDetails, so it doesn't get pulled into these tests
jest.mock('./EntryDetails');
require('./EntryDetails').default.mockImplementation(() => null);

describe('the Activities component', () => {
  const state = {
    apd: {
      data: {
        activities: ['activity 1', 'activity 2', 'activity 3']
      }
    }
  };

  const store = mockStore(state);

  beforeEach(() => {
    store.clearActions();
  });

  test('renders correctly', () => {
    const component = mount(
      <Provider store={store}>
        <Activities />
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
});
