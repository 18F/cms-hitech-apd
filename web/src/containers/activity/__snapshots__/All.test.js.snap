// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Activities component renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <All>
    <Connect(ConnectedWaypoint)
      id="activities-overview"
    >
      <ConnectedWaypoint
        id="activities-overview"
        scrollTo={[Function]}
      >
        <Waypoint
          bottomOffset="90%"
          debug={false}
          fireOnRapidScroll={true}
          horizontal={false}
          onEnter={[Function]}
          onLeave={[Function]}
          onPositionChange={[Function]}
          topOffset="0px"
        >
          <span
            style={
              Object {
                "fontSize": 0,
              }
            }
          />
        </Waypoint>
        <Section
          id="activities"
          isNumbered={true}
          resource="activities"
        >
          <section
            className="numbered-section"
            id="activities"
          >
            <h2
              className="ds-h2"
            >
              Program Activities
            </h2>
            <span
              className="ds-text--lead"
            >
              Explain the goals, timelines, and costs associated with each activity. Activities are narrowly scoped bodies of work that comprise the APD as described within the Program Overview section.
            </span>
            <Connect(ConnectedWaypoint)
              id="activities-list"
            >
              <ConnectedWaypoint
                id="activities-list"
                scrollTo={[Function]}
              >
                <Waypoint
                  bottomOffset="90%"
                  debug={false}
                  fireOnRapidScroll={true}
                  horizontal={false}
                  onEnter={[Function]}
                  onLeave={[Function]}
                  onPositionChange={[Function]}
                  topOffset="0px"
                >
                  <span
                    style={
                      Object {
                        "fontSize": 0,
                      }
                    }
                  />
                </Waypoint>
              </ConnectedWaypoint>
            </Connect(ConnectedWaypoint)>
            <Subsection
              headerClassName={null}
              id="activities-list"
              nested={false}
              open={true}
              resource="activities.list"
            >
              <h3
                className=" subsection--title ds-h3"
                id="activities-list"
              >
                Activity List
              </h3>
              <Instruction
                args={null}
                headingDisplay={
                  Object {
                    "className": "ds-h3",
                    "level": "h3",
                  }
                }
                reverse={false}
                source="activities.list.instruction"
              >
                <div>
                  <div
                    className="visibility--screen"
                  >
                    <p
                      className="ds-u-font-weight--bold"
                    >
                      List as many activities as necessary to cover all scope related to this APD.
                    </p>
                    <Md
                      content="Examples of individual activities include: auditing, environmental scans, HIE registry development, CQM database development, data and analytics, HIE onboarding, etc."
                      wrapper="p"
                    >
                      <p
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "Examples of individual activities include: auditing, environmental scans, HIE registry development, CQM database development, data and analytics, HIE onboarding, etc.",
                          }
                        }
                      />
                    </Md>
                    <Md
                      className="instruction-box"
                      content="Program administration pertains to implementing the EHR Incentive program itself, including state and contractor costs associated with operating EHR incentive program and SLR enhancements/upgrades. All APDs must provide activity details for program administration."
                      wrapper="p"
                    >
                      <p
                        className="instruction-box"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "Program administration pertains to implementing the EHR Incentive program itself, including state and contractor costs associated with operating EHR incentive program and SLR enhancements/upgrades. All APDs must provide activity details for program administration.",
                          }
                        }
                      />
                    </Md>
                  </div>
                </div>
              </Instruction>
              <NameAndFundingSourceReview
                disableExpand={true}
                expand={[Function]}
                index={-1}
                item="activity 1"
                onDeleteClick={null}
              >
                <StandardReview
                  editHref={null}
                  heading="1. undefined"
                  headingLevel={4}
                  onDeleteClick={null}
                  onEditClick={null}
                >
                  <Review
                    editContent={
                      <div
                        className="nowrap visibility--screen"
                      />
                    }
                    editText="Edit"
                    heading="1. undefined"
                    headingLevel={4}
                  >
                    <div
                      className="ds-c-review ds-u-border-bottom--2 ds-u-padding-y--2 ds-u-justify-content--between ds-u-display--flex"
                    >
                      <div
                        className="ds-u-margin-right--2"
                      >
                        <h4
                          className="ds-c-review__heading ds-text ds-u-margin-bottom--0 ds-u-font-weight--bold ds-u-display--inline-block"
                        >
                          1. undefined
                        </h4>
                        <div
                          className="ds-c-review__body"
                        >
                          <strong>
                            Program type:
                          </strong>
                           
                        </div>
                      </div>
                      <div
                        className="nowrap visibility--screen"
                      />
                    </div>
                  </Review>
                </StandardReview>
              </NameAndFundingSourceReview>
              <FormAndReviewList
                addButtonText="Add another activity"
                allowDeleteAll={true}
                className="ds-u-border-bottom--0"
                collapsed={[Function]}
                expanded={[Function]}
                handleChange={[Function]}
                list={
                  Array [
                    "activity 2",
                    "activity 3",
                  ]
                }
                noDataMessage={false}
                onAddClick={[Function]}
                onDeleteClick={[Function]}
              >
                <div
                  className="form-and-review-list ds-u-border-bottom--0"
                >
                  <FormAndReviewItem
                    collapsedComponent={[Function]}
                    expandedComponent={[Function]}
                    handleChange={[Function]}
                    index={0}
                    initialCollapsed={true}
                    item="activity 2"
                    onDeleteClick={[Function]}
                  >
                    <div
                      className="form-and-review-list--item__collapsed"
                    >
                      <NameAndFundingSourceReview
                        disableExpand={false}
                        expand={[Function]}
                        handleChange={[Function]}
                        index={0}
                        item="activity 2"
                        onDeleteClick={[Function]}
                      >
                        <StandardReview
                          editHref={null}
                          heading="2. undefined"
                          headingLevel={4}
                          onDeleteClick={[Function]}
                          onEditClick={[Function]}
                        >
                          <Review
                            editContent={
                              <div
                                className="nowrap visibility--screen"
                              >
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  Edit
                                </Button>
                                <React.Fragment>
                                  |
                                  <Button
                                    onClick={[Function]}
                                    size="small"
                                    type="button"
                                    variation="transparent"
                                  >
                                    Remove
                                  </Button>
                                </React.Fragment>
                              </div>
                            }
                            editText="Edit"
                            heading="2. undefined"
                            headingLevel={4}
                          >
                            <div
                              className="ds-c-review ds-u-border-bottom--2 ds-u-padding-y--2 ds-u-justify-content--between ds-u-display--flex"
                            >
                              <div
                                className="ds-u-margin-right--2"
                              >
                                <h4
                                  className="ds-c-review__heading ds-text ds-u-margin-bottom--0 ds-u-font-weight--bold ds-u-display--inline-block"
                                >
                                  2. undefined
                                </h4>
                                <div
                                  className="ds-c-review__body"
                                >
                                  <strong>
                                    Program type:
                                  </strong>
                                   
                                </div>
                              </div>
                              <div
                                className="nowrap visibility--screen"
                              >
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  <button
                                    className="ds-c-button ds-c-button--transparent ds-c-button--small"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Edit
                                  </button>
                                </Button>
                                |
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  <button
                                    className="ds-c-button ds-c-button--transparent ds-c-button--small"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Remove
                                  </button>
                                </Button>
                              </div>
                            </div>
                          </Review>
                        </StandardReview>
                      </NameAndFundingSourceReview>
                    </div>
                  </FormAndReviewItem>
                  <FormAndReviewItem
                    collapsedComponent={[Function]}
                    expandedComponent={[Function]}
                    handleChange={[Function]}
                    index={1}
                    initialCollapsed={true}
                    item="activity 3"
                    onDeleteClick={[Function]}
                  >
                    <div
                      className="form-and-review-list--item__collapsed"
                    >
                      <NameAndFundingSourceReview
                        disableExpand={false}
                        expand={[Function]}
                        handleChange={[Function]}
                        index={1}
                        item="activity 3"
                        onDeleteClick={[Function]}
                      >
                        <StandardReview
                          editHref={null}
                          heading="3. undefined"
                          headingLevel={4}
                          onDeleteClick={[Function]}
                          onEditClick={[Function]}
                        >
                          <Review
                            editContent={
                              <div
                                className="nowrap visibility--screen"
                              >
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  Edit
                                </Button>
                                <React.Fragment>
                                  |
                                  <Button
                                    onClick={[Function]}
                                    size="small"
                                    type="button"
                                    variation="transparent"
                                  >
                                    Remove
                                  </Button>
                                </React.Fragment>
                              </div>
                            }
                            editText="Edit"
                            heading="3. undefined"
                            headingLevel={4}
                          >
                            <div
                              className="ds-c-review ds-u-border-bottom--2 ds-u-padding-y--2 ds-u-justify-content--between ds-u-display--flex"
                            >
                              <div
                                className="ds-u-margin-right--2"
                              >
                                <h4
                                  className="ds-c-review__heading ds-text ds-u-margin-bottom--0 ds-u-font-weight--bold ds-u-display--inline-block"
                                >
                                  3. undefined
                                </h4>
                                <div
                                  className="ds-c-review__body"
                                >
                                  <strong>
                                    Program type:
                                  </strong>
                                   
                                </div>
                              </div>
                              <div
                                className="nowrap visibility--screen"
                              >
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  <button
                                    className="ds-c-button ds-c-button--transparent ds-c-button--small"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Edit
                                  </button>
                                </Button>
                                |
                                <Button
                                  onClick={[Function]}
                                  size="small"
                                  type="button"
                                  variation="transparent"
                                >
                                  <button
                                    className="ds-c-button ds-c-button--transparent ds-c-button--small"
                                    onClick={[Function]}
                                    type="button"
                                  >
                                    Remove
                                  </button>
                                </Button>
                              </div>
                            </div>
                          </Review>
                        </StandardReview>
                      </NameAndFundingSourceReview>
                    </div>
                  </FormAndReviewItem>
                  <Button
                    className="visibility--screen"
                    onClick={[Function]}
                    type="button"
                  >
                    <button
                      className="ds-c-button visibility--screen"
                      onClick={[Function]}
                      type="button"
                    >
                      Add another activity
                    </button>
                  </Button>
                </div>
              </FormAndReviewList>
            </Subsection>
            <Connect(ConnectedWaypoint)>
              <ConnectedWaypoint
                scrollTo={[Function]}
              >
                <Waypoint
                  bottomOffset="90%"
                  debug={false}
                  fireOnRapidScroll={true}
                  horizontal={false}
                  onEnter={[Function]}
                  onLeave={[Function]}
                  onPositionChange={[Function]}
                  topOffset="0px"
                >
                  <span
                    style={
                      Object {
                        "fontSize": 0,
                      }
                    }
                  />
                </Waypoint>
                <EntryDetails
                  activity="activity 1"
                  index={0}
                  num={1}
                />
              </ConnectedWaypoint>
            </Connect(ConnectedWaypoint)>
            <Connect(ConnectedWaypoint)>
              <ConnectedWaypoint
                scrollTo={[Function]}
              >
                <Waypoint
                  bottomOffset="90%"
                  debug={false}
                  fireOnRapidScroll={true}
                  horizontal={false}
                  onEnter={[Function]}
                  onLeave={[Function]}
                  onPositionChange={[Function]}
                  topOffset="0px"
                >
                  <span
                    style={
                      Object {
                        "fontSize": 0,
                      }
                    }
                  />
                </Waypoint>
                <EntryDetails
                  activity="activity 2"
                  index={1}
                  num={2}
                />
              </ConnectedWaypoint>
            </Connect(ConnectedWaypoint)>
            <Connect(ConnectedWaypoint)>
              <ConnectedWaypoint
                scrollTo={[Function]}
              >
                <Waypoint
                  bottomOffset="90%"
                  debug={false}
                  fireOnRapidScroll={true}
                  horizontal={false}
                  onEnter={[Function]}
                  onLeave={[Function]}
                  onPositionChange={[Function]}
                  topOffset="0px"
                >
                  <span
                    style={
                      Object {
                        "fontSize": 0,
                      }
                    }
                  />
                </Waypoint>
                <EntryDetails
                  activity="activity 3"
                  index={2}
                  num={3}
                />
              </ConnectedWaypoint>
            </Connect(ConnectedWaypoint)>
          </section>
        </Section>
      </ConnectedWaypoint>
    </Connect(ConnectedWaypoint)>
  </All>
</Provider>
`;
